File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Is Graph Bipartite?/Java/dfs/Solution.java
import java.util.*;

class Solution {
    public boolean isBipartite(int[][] graph) {
        int[] colors = new int[graph.length];
        boolean[] visited = new boolean[graph.length];
        for (int start = 0; start < graph.length; start++) {
            if (visited[start]) {
                continue;
            }

            if (!dfs(graph, start, 1, colors, visited)) {
                return false;
            }
        }

        return true;
    }

    private boolean dfs(int[][] graph, int curV, int curColor, int[] colors, boolean[] visited) {
        visited[curV] = true;
        colors[curV] = curColor;
        for (int nextV : graph[curV]) {
            if (!visited[nextV]) {
                if (!dfs(graph, nextV, 3 - curColor, colors, visited)) {
                    return false;
                };
            }

            if (colors[nextV] == curColor) {
                return false;
            }
        }   

        return true;
    }
}

File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Is Graph Bipartite?/dfs/solution.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        visited = [False for _ in graph]
        node_color = [None for _ in graph]

        for start in range(len(graph)):
            if visited[start]:
                continue

            if not self.dfs(graph, visited, node_color, [start, 'red']):
                return False
        
        return True

    def dfs(self, graph, visited, node_color, current):
        visited[current[0]] = True
        node_color[current[0]] = current[1]

        next_color = 'black' if current[1] == 'red' else 'red'
        for next in graph[current[0]]:
            if visited[next]:
                if node_color[next] != next_color:
                    return False
                else:
                    continue
            
            if not self.dfs(graph, visited, node_color, [next, next_color]):
                return False
        
        return True
